mobile/SpaceLaunchNews/app/src/main/java/com/wojbeg/spacelaunchnews/ui/fragments/LaunchNewsFragment.kt

```diff
@@ -0,0 +1,155 @@
+package com.wojbeg.spacelaunchnews.ui.fragments
+
+import android.os.Bundle
+import android.view.View
+import android.widget.AbsListView
+import android.widget.Toast
+import androidx.fragment.app.Fragment
+import androidx.fragment.app.activityViewModels
+import androidx.lifecycle.Observer
+import androidx.navigation.fragment.findNavController
+import androidx.recyclerview.widget.LinearLayoutManager
+import androidx.recyclerview.widget.RecyclerView
+import com.wojbeg.spacelaunchnews.R
+import com.wojbeg.spacelaunchnews.adapters.ArticleAdapter
+import com.wojbeg.spacelaunchnews.databinding.FragmentLaunchNewsBinding
+import com.wojbeg.spacelaunchnews.ui.viewmodels.NewsViewModel
+import com.wojbeg.spacelaunchnews.util.Constants
+import com.wojbeg.spacelaunchnews.util.Resource
+import dagger.hilt.android.AndroidEntryPoint
+import javax.inject.Inject
+
+@AndroidEntryPoint
+class LaunchNewsFragment : Fragment(R.layout.fragment_launch_news) {
+
+    private val viewModel: NewsViewModel by activityViewModels()
+    private lateinit var binding: FragmentLaunchNewsBinding
+
+    @Inject
+    lateinit var articleAdapter: ArticleAdapter
+
+    var isLoading = false
+    var isLastPage = false
+    var isScrolling = false
+
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        binding = FragmentLaunchNewsBinding.bind(view)
+
+        setupRecyclerView()
+
+        articleAdapter.setOnItemClickListener {
+            val bundle = Bundle().apply {
+                putSerializable("article", it)
+            }
+
+            findNavController().navigate(
+                R.id.action_launchNewsFragment_to_articleFragment,
+                bundle
+            )
+        }
+
+
+        viewModel.reportsLiveData.observe(viewLifecycleOwner, Observer {
+            response ->
+            when(response) {
+                is Resource.Success ->{
+                    hideProgressBar()
+                    response.data?.let {
+                        articleResponse ->
+                        articleAdapter.differ.submitList(articleResponse)
+
+                        articleAdapter.notifyDataSetChanged()
+
+                        isLastPage = viewModel.isLastPage()
+
+                        if(isLastPage){
+                            binding.rvBreakingNews.setPadding(0,0,0,0)
+                        }
+
+                    }
+                }
+
+                is Resource.Error ->{
+                    hideProgressBar()
+                    response.message?.let {
+                        Toast.makeText(activity, getString(R.string.error) + it, Toast.LENGTH_LONG).show()
+                    }
+                }
+                is Resource.Loading ->{
+                    showProgressBar()
+                }
```

Comment:
Nie jest to błąd, ale ja osobiście lubię zrobić sobie odrębne metody dla obsługi poszczególnych stanów widoku np. `fun handleSuccess(result: List<coś tam>)`. Wtedy taka metoda jak ta może wyglądać w taki zwięzły i krótki sposób:
 ```suggestion
                when(response) {
                    is Response.Success -> handleSuccess(it)
                    is Response.Loading -> handleLoading()
                    is Response.Error -> handleError(it)
                }
```